Implement killing of subprocesses

M-Patrone opened the pull request:
Fixes #43

M-Patrone commented:
Not sure why the test fails, locally it runs (Windows 10)

M-Patrone commented:
Another thing to note is that killing subprocesses does not work on .net standard 2.0 as there is no API for doing this. The [best API](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.kill?view=netcore-3.0#System_Diagnostics_Process_Kill_System_Boolean_ ) is the one in .Net Core 3.0 which is finally crossplat

Tyrrrz commented:
Thanks for the PR @M-Patrone.
I have only glanced through it for now, but I have a suggestion: let's not add an option `killProcessTree`, rather make it implicit. This will avoid some `#if/#endif` directives. `KillProcessAndChildrens` can just no-op if it's not supported.

Regarding CI build, if I remember correctly, tests are targetting `.net45` which is an AppVeyor limitation (it doesn't extract test results from .net core tests), so maybe that's why the tests fail.

M-Patrone commented:
Tanks for quick response @Tyrrrz. I'm not 100% sure if I really understood you correct, but I removed the property and the optional parameter in the method to kill the entire process tree. If we can, we are going to kill the entire process tree with this pull. Again, in the locally environment, the Test "Execute_Sleep_InSubprocess_CancelLate_Test" runs successfully.

Tyrrrz commented on CliWrap.Tests/CliTests.cs:
```diff
@@ -131,6 +133,40 @@ public void Execute_EchoSpam_Callback_Test()
+                });
+            }
+        }
+
```
These tests don't seem to be making sure that child processes are killed. In fact, `OperationCanceledException` will be thrown either way. It appears that these tests don't add value on top of the previously existing `Execute_Sleep_CancelEarly_Test` tests.

Tyrrrz commented on CliWrap.Tests/CliTests.cs:
```diff
@@ -318,6 +354,41 @@ public async Task ExecuteAsync_EchoSpam_Callback_Test()
+                });
+            }
+        }
+
```
Same story here

Tyrrrz commented:
>Isn't this quite breaking? Not sure if this is expected by all users

I'd probably expect it to be the original behavior and thus this being a bug fix, but you're right, this might not be the same for other users.

M-Patrone commented on CliWrap/Internal/CliProcess.cs:
```diff
@@ -183,5 +192,31 @@ public void Dispose()
             _standardErrorEndSignal.Dispose();
         }
+#if NET45
+        private static void KillProcessAndChildrens(int pid)
```
As it's recursive, I need the parameter

Tyrrrz replied:
Ah, didn't notice, sorry.

M-Patrone commented:
Thanks @Tyrrrz  for your help and sorry for my late response. I have now implemented the suggested improvements:
- Tests: Check if child processes are terminated
- rename: KillProcessAndChildrens -->KillProcessTree
- removed: TargetFramework .NET Core 3.0

Tyrrrz commented:
Hi.
Can you please allow committing to your branch (there should be an option in the right section of the PR)? I will commit some small changes and merge.

M-Patrone commented:
Hi @Tyrrrz 
I think I did it, but I am not sure if thats correct. 
![image](https://user-images.githubusercontent.com/14307347/67841458-6384ed00-faf8-11e9-8f65-205adda12370.png)

Tyrrrz commented:
Looks like I still can't push, probably because the repository owned by someone else. It's okay, I'll merge and change it locally.

M-Patrone commented:
Hi @Tyrrrz,  thank you very much. Sorry for that!
